Conditional get rework

crankycoder opened the pull request:
This is an addendum to #1398. 

This removes the isTileCurrent check and just moves it together with the actual tile download.

The significant change over @cascheberg's original patch is that the TileIOFacade will now return a SerializableTile after saving the tile.  This can then be immediately used to load an instance of Drawable.

An HTTP 304 will also check that the tile can be properly loaded from disk in the case that the tile was somehow corrupted on disk.  If this is the case, the tile's etag is completely invalidated and the returned tile is set to NULL.

crankycoder committed:
Refactoring inner classes out to module level to ease testing.

cascheberg committed:
improve etag usage when downloading map tiles

crankycoder committed:
reworked the code to simplify codepath for file downloads

crankycoder committed:
More refactoring to inner classes to module level classes.

crankycoder commented:
This still needs a testcase.

crankycoder committed:
Renamed SmartFSAbstractTileLoader to SmartFSTileLoader as it's not
actually an abstract class.

crankycoder committed:
added test fixtures for tile load tests

crankycoder committed:
Renamed httpResponse to httpStatusCode()

crankycoder committed:
Pulled SmartFSTileLoader to the module level to make testing easier.
Moved MockHttpUtil into the test branch of the source from the actual application code.
Added a basic testcase that runs the TileDownloaderDelegate::downloadTile method to completion

cascheberg commented on android/src/main/java/org/mozilla/osmdroid/tileprovider/modules/TileDownloaderDelegate.java:
```diff
@@ -52,80 +49,31 @@ public TileDownloaderDelegate(INetworkAvailablityCheck pNetworkAvailablityCheck,
-
-        if (urlIs404Cached(tileURLString)) {
-            return false;
+            return tileSource.getDrawable(serializableTile.getTileData());
```
@crankycoder Almost all `return null` in this method should be replaced by returning the cached tile, similar to this. Otherwise there is no map if offline etc.

crankycoder replied:
I'd rather see that added in as a separate class and implemented as a proxy
to the delegate.

Separate the concerns and you can easily test that.  Load the image from
the tiledownloader,  if that fails,  try to load from disk.
On Jan 17, 2015 4:23 AM, "Christian Ascheberg" notifications@github.com
wrote:

> In
> android/src/main/java/org/mozilla/osmdroid/tileprovider/modules/TileDownloaderDelegate.java
> https://github.com/mozilla/MozStumbler/pull/1400#discussion-diff-23124047
> :
> 
> > -            return false;
> > 
> > ## \-        }
> > -        if (networkIsUnavailable()) {
> > -            return false;
> > 
> > ## \-        }
> > 
> > ## \-        final String tileURLString = tileSource.getTileURLString(tile);
> > -        if (tileURLString == null || tileURLString.length() == 0) {
> > -            return false;
> > 
> > ## \-        }
> > -        if (urlIs404Cached(tileURLString)) {
> > -            return false;
> > -            return tileSource.getDrawable(serializableTile.getTileData());
> 
> @crankycoder https://github.com/crankycoder Almost all return null in
> this method should be replaced by returning the cached tile, similar to
> this. Otherwise there is no map if offline etc.
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/mozilla/MozStumbler/pull/1400/files#r23124047.

crankycoder committed:
added basic test case demonstrating how to load a test fixture

crankycoder committed:
Updated test.sh script and the Groovy plugin to invoke copyTestResources prior to running tests.
Updated Makefile to invoke the copyTestResources task when `make test`
Blargh.  Slow.

crankycoder committed:
Updated TileDownloaderDelegateTest so that tile fixture data is properly loaded from disk.

crankycoder committed:
replaced real downloading with mock downloads

crankycoder committed:
added tests for 3 basic cases:

* simple HTTP200 download
* HTTP 304 conditional GET with existing ondisk cache
* HTTP 304 conditional GET with corrupt on disk cache

crankycoder committed:
added 2 tests for network failure.
* Tiles are returned if ondisk tiles already exist
* NULL is returned if no ondisk tile already exists

crankycoder commented:
@garvankeeley ok, tests are in.  You'll find tests that cover the following cases:
- basic HTTP200
- HTTP 304 conditional GET with on disk tile 
- HTTP 304 conditional GET with a tile that has zero bytes - which should result in a NULL instead of a tile
- Network outage should return tile if it exists on disk
- Network outage should return null if no tile exists on disk

crankycoder commented:
One more note - I had to modify the build because gradle and android are terrible.  3461ff7 includes patches to android/build.gradle and the Groovy test runner that will manually copy over test resources so that you can use test fixture data.  

![rocket_racoon](https://cloud.githubusercontent.com/assets/4829/5812013/9eec46c4-a02e-11e4-9f88-39fdffa09903.png)

garvankeeley commented on android/src/main/java/org/mozilla/osmdroid/tileprovider/modules/SmartFSTileLoader.java:
```diff
@@ -0,0 +1,84 @@
+        final MapTile tile = pState.getMapTile();
+
+        // if there's no sdcard then don't do anything
+        if (!smartFSProvider.getSdCardAvailable()) {
```
We don't depend on an SD card.

crankycoder replied:
That's the same code that was in the TileLoader before.  I just moved the code and didn't want to add any functional changes in this patch.
